Add kopf create

rodrigofronza opened the pull request:
Add kopf create

Issue: #11

rodrigofronza committed:
Add kopf create

Issue: #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
Add build container function

Issue: #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
Add Dockerfile for operator

Issue: #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
Add the Kubernetes operator to deploy the operator

Issue: #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
refactor(operator): replace operator placeholders

Replace operator placeholders with actual values.

Closes #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

NunoFrRibeiro commented on Dockerfile:
Not really needed anymore, since it will be replace by the dagger build container function

NunoFrRibeiro commented on .dagger/src/repermon/main.py:
```diff
@@ -78,3 +78,17 @@ def __test_manifests(self) -> str:
+            .with_exec(["sh", "-c", "uv sync --project repermon"])
+            .with_exec(["sh", "-c", "source .venv/bin/activate"])
+            .with_entrypoint(["kopf", "run", "/app/src/operator.py"])
+        )
```
This seems like a good first pass at the build container function, but it can be optimized quite a bit and we should really try and follow dagger motto of reusability and modularity.

One approach could be, for example, create inside the https://github.com/nukamigo/repermon/blob/main/.dagger/src/repermon/ci.py file a new class for build functions (build/test/lint/push).

so, let's a have go at this approach?

NunoFrRibeiro commented on kubernetes/operator.yaml:
```diff
@@ -0,0 +1,76 @@
+kind: ServiceAccount
+metadata:
+  name: my-operator
+  namespace: default
```
don't forget to change all namespaces to `repermon` or something along those ways

also, what should we call this operator?
- repermon-operator?
- ...

rodrigofronza committed:
feat: add build container function

Add a new build container function to have a generic function for it.

Resolves #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
fix(dagger): fix pre commit

Fix the pre-commit function by making sure that git is installed.

Resolves #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

rodrigofronza committed:
fix(dagger): fix test cluster fuction

Fix the test cluster function by applying each manifest sequentially.

Resolves #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>

NunoFrRibeiro commented on .dagger/src/repermon/main.py:
```diff
@@ -75,6 +75,8 @@ def __test_manifests(self) -> str:
-            .with_exec(["kubectl", "apply", "-f", "/manifests"])
+            .with_exec(["kubectl", "apply", "-f", "/manifests/crd.yaml"])
+            .with_exec(["kubectl", "apply", "-f", "/manifests/rbac.yaml"])
+            .with_exec(["kubectl", "apply", "-f", "/manifests/operator.yaml"])
```
you can just apply the whole folder of the manifests
`kubectl apply -f ./manifests/`

but while I was testing calling the function with:
`dagger call test-cluster` I found  mistake we were forgeting.

The name space `ai-monitor` needs to exist or we need to create it

rodrigofronza replied:
Exactly, that's why I've executed each manifest sequentially, to create the namespace.

NunoFrRibeiro commented on kubernetes/operator.yaml:
```diff
@@ -0,0 +1,76 @@
+apiVersion: v1
```
before creating the `ServiceAccount` can you add the manifest for creating the name space?
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: ai-monitor
```

rodrigofronza replied:
Absolutely, that makes sense

NunoFrRibeiro commented on .dagger/src/repermon/ci.py:
```diff
@@ -13,13 +30,41 @@ def __init__(self, source: dagger.Directory) -> None:
-            .with_exec(["sh", "-c", "uv sync --project repermon"])
-            .with_exec(["sh", "-c", "source .venv/bin/activate"])
-            .with_exec(["sh", "-c", ".venv/bin/pre-commit run --all-files"])
+            self.build_container("ghcr.io/astral-sh/uv:python3.12-alpine", self.source)
```
I think the docker image could be moved into the `BuildContainer` class
Or if we want to maintain it like this have be an `Optional` argument for the pch function and have a default in the `BuildContainer` class

rodrigofronza replied:
I would say to have it as an optional, because the intention of the `build_container` function is to be generic and build a container from any image.

rodrigofronza committed:
chore(dagger): make image opitional

- Make the image parameter of the `build_container` funcion as optional and also default to "ghcr.io/astral-sh/uv:python3.12-alpine".
- Create the namespace `ai-monitor` on the operator manifest.

Resolves #11

Signed-off-by: rodrigofronza <rodrigofronza2@gmail.com>
